lib/collection/src/collection.rs

```diff
@@ -258,17 +258,87 @@ impl Collection {
         todo!("Handle for sender and receiver")
     }
 
+    /// Initiate temporary shard
+    ///
+    /// Drops existing temporary shards for `shard_id`.
+    pub async fn initiate_temporary_shard(&self, shard_id: ShardId) -> CollectionResult<()> {
+        let mut shard_holder_write = self.shards_holder.write().await;
+        let temporary_shard_path = self.path.join(format!("{shard_id}-temp"));
+
+        if let Some(mut existing_temporary_shard) =
+            shard_holder_write.remove_temporary_shard(shard_id)
+        {
+            log::info!(
+                "A temporary shard is already present for {}:{} - its content will be deleted",
+                self.id,
+                shard_id
+            );
+            // Finish update tasks
+            existing_temporary_shard.before_drop().await;
+            // Delete existing folder
+            tokio::fs::remove_dir_all(&temporary_shard_path).await?;
+        }
+        // do not lock shards while creating temporary shard  disk
+        drop(shard_holder_write);
+
+        // create directory to hold the temporary data for shard with `shard_id`
+        tokio::fs::create_dir_all(&temporary_shard_path).await?;
+        let temporary_shard = LocalShard::build(
+            shard_id,
+            self.id.clone(),
+            &temporary_shard_path,
+            self.config.clone(),
+        )
+        .await?;
+
+        // register temporary shard
+        let mut shards_holder_write = self.shards_holder.write().await;
+        shards_holder_write.add_temporary_shard(shard_id, temporary_shard);
+        Ok(())
+    }
+
+    /// Handle collection updates from peers.
+    ///
+    /// Shard transfer aware.
     pub async fn update_from_peer(
         &self,
         operation: CollectionUpdateOperations,
         shard_selection: ShardId,
         wait: bool,
     ) -> CollectionResult<UpdateResult> {
-        let local_shard = self
-            .shards_holder
-            .local_shard_by_id(shard_selection)
-            .await?;
-        local_shard.get().update(operation.clone(), wait).await
+        let shard_holder_guard = self.shards_holder.read().await;
+        let shard_opt = shard_holder_guard.get_shard(&shard_selection);
+
+        let target_shard = match shard_opt {
+            None => {
```

Comment:
Does a temporary shard have a new id? That's why we can get `None` here?

Reply:
The temporary shard should get the same id.

We have decided to remove much of the restrictions here to make the nodes more reliable to being out of sync.

Reply:
Ok, got it